File: DemoApp Objective-C/CASSample/SupportFiles/Coordinator/AppCoordinator.m
//
//  AppCoordinator.m
//  CASSample
//

#import "AppCoordinator.h"
#import "MainVC.h"
#import "BannerVC.h"
#import "InterstitialVC.h"
#import "RewardedVC.h"
#import "AppOpenVC.h"
#import "NativeVC.h"
#import "NativeTemplateVC.h"

@interface AppCoordinator ()

@property (nonatomic, strong) UIWindow *window;
@property (nonatomic, strong) UIStoryboard *storyboard;
@property (nonatomic, strong) UINavigationController *navigationController;

@end

@implementation AppCoordinator

- (instancetype)initWithWindow:(UIWindow *)window {
    self = [super init];
    if (self) {
        _window = window;
        _storyboard = [UIStoryboard storyboardWithName:@"Main" bundle:nil];
    }
    return self;
}

- (void)start {
    MainVC *mainVC = (MainVC *)[self.storyboard instantiateViewControllerWithIdentifier:@"MainVC"];
    mainVC.coordinator = self;

    self.navigationController = [[UINavigationController alloc] initWithRootViewController:mainVC];
    self.navigationController.navigationBar.prefersLargeTitles = YES;
    
    self.window.rootViewController = self.navigationController;
    [self.window makeKeyAndVisible];
}

- (void)navigateToBannerAd {
    BannerVC *vc = (BannerVC *)[self.storyboard instantiateViewControllerWithIdentifier:@"BannerVC"];
    [self.navigationController pushViewController:vc animated:YES];
}

- (void)navigateToNativeAd {
    NativeVC *vc = (NativeVC *)[self.storyboard instantiateViewControllerWithIdentifier:@"NativeVC"];
    [self.navigationController pushViewController:vc animated:YES];
}

- (void)navigateToNativeTemplateAd {
    NativeTemplateVC *vc = (NativeTemplateVC *)[self.storyboard instantiateViewControllerWithIdentifier:@"NativeTemplateVC"];
    [self.navigationController pushViewController:vc animated:YES];
}

- (void)navigateToInterstitialAd {
    InterstitialVC *vc = (InterstitialVC *)[self.storyboard instantiateViewControllerWithIdentifier:@"InterstitialVC"];
    [self.navigationController pushViewController:vc animated:YES];
}

- (void)navigateToRewardedAd {
    RewardedVC *vc = (RewardedVC *)[self.storyboard instantiateViewControllerWithIdentifier:@"RewardedVC"];
    [self.navigationController pushViewController:vc animated:YES];
}

- (void)navigateToAppOpenAd {
    AppOpenVC *vc = (AppOpenVC *)[self.storyboard instantiateViewControllerWithIdentifier:@"AppOpenVC"];
    [self.navigationController pushViewController:vc animated:YES];
}

- (void)goBack {
    [self.navigationController popViewControllerAnimated:YES];
}

@end

File: DemoApp Swift/CASSample/SupportFiles/Coordinator/AppCoordinator.swift
//
//  AppCoordinator.swift
//  CASSample
//
//  Copyright © 2025 Clever Ads Solutions. All rights reserved.
//

import UIKit
import CleverAdsSolutions

class AppCoordinator {
    
    private let window: UIWindow
    private let storyboard: UIStoryboard
    
    private var navigationController: UINavigationController!
    
    init(window: UIWindow) {
        self.window = window
        self.storyboard = UIStoryboard(name: "Main", bundle: nil)
    }

    func start() {
        let mainVC = instantiateMainMenuViewController()
        navigationController = UINavigationController(rootViewController: mainVC)
        navigationController.navigationBar.prefersLargeTitles = true
        window.rootViewController = navigationController
        window.makeKeyAndVisible()
    }
    
    private func instantiateMainMenuViewController() -> MainVC {
        let vc = storyboard.instantiateViewController(withIdentifier: "MainVC") as! MainVC
        vc.coordinator = self
        return vc
    }
    
    func navigateToBannerAd() {
        let vc = storyboard.instantiateViewController(withIdentifier: "BannerVC") as! BannerVC
        navigationController.pushViewController(vc, animated: true)
    }
    
    func navigateToNativeAd() {
        let vc = storyboard.instantiateViewController(withIdentifier: "NativeVC") as! NativeVC
        navigationController.pushViewController(vc, animated: true)
    }
    
    func navigateToNativeTemplate() {
        let vc = storyboard.instantiateViewController(withIdentifier: "NativeTemplateVC") as! NativeTemplateVC
        navigationController.pushViewController(vc, animated: true)
    }
    
    func navigateToInterstitialAd() {       
        let vc = storyboard.instantiateViewController(withIdentifier: "InterstitialVC") as! InterstitialVC
        navigationController.pushViewController(vc, animated: true)
    }
    
    func navigateToRewardedAd() {
        let vc = storyboard.instantiateViewController(withIdentifier: "RewardedVC") as! RewardedVC
        navigationController.pushViewController(vc, animated: true)
    }
    
    func navigateToAppOpenAd() {
        let vc = storyboard.instantiateViewController(withIdentifier: "AppOpenVC") as! AppOpenVC        
        navigationController.pushViewController(vc, animated: true)
    }
       
    func goBack() {
        navigationController.popViewController(animated: true)
    }
}
